### PR TITLE
Torii: Remove hex strings leading zeros from GraphQL responses

### PR DESCRIPTION
**Pull request by akhercha**
Resolves: #1008

### CODE REVIEW

File: crates/dojo-types/src/primitive.rs [hunk 1]
```diff
@@ -70,7 +70,15 @@ impl Primitive {
...
             | Primitive::ClassHash(_)
-            | Primitive::Felt252(_) => Ok(format!("'0x{:064x}'", value[0])),
+            | Primitive::Felt252(_) => {
+                let hex_string = format!("{:x}", value[0]);
+                let trimmed_hex = hex_string.trim_start_matches('0');
+                Ok(if trimmed_hex.is_empty() {
+                    "'0x0'".to_string()
+                } else {
+                    format!("'0x{}'", trimmed_hex)
+                })
```
Comment (tarrencev) [inline crates/dojo-types/src/primitive.rs:80]: I believe we want to continue storing the prefixed hex values (for lexical sorting) but want to remove the prefix in the graphql api before returning it to the user

Comment (broody) [reply crates/dojo-types/src/primitive.rs:80]: Right, should happen on graphql side

@tarrencev can we finish reviewing this https://github.com/dojoengine/dojo/pull/1005 and merge so this fix can be built on top?

Comment (akhercha) [reply crates/dojo-types/src/primitive.rs]: Ah right, my baaaad.
Should be it now, let me know if I can do better!

Comment (broody) [reply crates/dojo-types/src/primitive.rs]: Works great!

image

File: crates/torii/graphql/src/query/mod.rs [hunk 2]
```diff
@@ -95,6 +95,16 @@ fn parse_nested_type(
     TypeData::Nested((TypeRef::named(target_type), nested_mapping))
 }
 
+fn remove_trailing_zeros(value: Value) -> Value {
```
Comment (broody) [inline crates/torii/graphql/src/query/mod.rs:98]: I think this should be `remove_leading_zeros`?

Comment (akhercha) [reply crates/torii/graphql/src/query/mod.rs]: Oops, updated!

File: crates/torii/graphql/src/query/mod.rs [hunk 3]
```diff
@@ -95,6 +95,16 @@ fn parse_nested_type(
...
+    if let Value::String(str_val) = &value {
+        let hex_part = str_val.trim_start_matches("0x");
```
Comment (tarrencev) [inline crates/torii/graphql/src/query/mod.rs:100]: Let's ensure that the string starts with `0x` before stripping

Comment (akhercha) [reply crates/torii/graphql/src/query/mod.rs:103]: Added!

File: crates/torii/graphql/src/query/mod.rs [hunk 4]
```diff
@@ -110,10 +123,20 @@ pub fn value_mapping_from_row(
...
-            ))
+            let field_type = type_data.type_ref().to_string();
+            let value = fetch_value(row, &column_name, &field_type)?;
+            let value = match Primitive::from_str(&field_type) {
+                Ok(primitive) => match primitive {
+                    Primitive::U128(_)
+                    | Primitive::ContractAddress(_)
+                    | Primitive::ClassHash(_)
+                    | Primitive::Felt252(_) => remove_hex_leading_zeros(value),
+                    _ => value,
```
Comment (tarrencev) [inline crates/torii/graphql/src/query/mod.rs:134]: Do we need U256 here too?

**Review by tarrencev — APPROVED** [review]
lgtm!

File: crates/torii/graphql/src/query/mod.rs (hunk 4 above)
Comment (akhercha) [reply crates/torii/graphql/src/query/mod.rs:134]: Ah yes; nice catch!
The function changed a little to handle this cause for this type the hexadecimal string does not start with 0x:
```rust
fn remove_hex_leading_zeros(value: Value) -> Value {
    if let Value::String(mut str_val) = value {
        if !str_val.starts_with("0x") {
            str_val = str_val.trim_start_matches("0x").to_string();
        }
        let trimmed_hex = str_val.trim_start_matches('0');
        Value::String(format!("0x{:0>1}", trimmed_hex))
    } else {
        value
    }
}
```

Comment (tarrencev) [reply crates/torii/graphql/src/query/mod.rs:134]: Hmm the hex string should always start with 0x. If that is not the case, we should fix it elsewhere

Comment (akhercha) [reply crates/torii/graphql/src/query/mod.rs:134]: Oh, thought it was intended; I can just add 0x here then:
```rust
Primitive::U256(_) => {
                if value.len() < 2 {
                    Err(PrimitiveError::NotEnoughFieldElements)
                } else {
                    let mut buffer = [0u8; 32];
                    let value0_bytes = value[0].to_bytes_be();
                    let value1_bytes = value[1].to_bytes_be();
                    buffer[..16].copy_from_slice(&value0_bytes);
                    buffer[16..].copy_from_slice(&value1_bytes);
                    Ok(format!("'{}'", hex::encode(buffer)))
                }
            }
```

Comment (akhercha) [reply crates/torii/graphql/src/query/mod.rs:134]: Does it make sense to have this constant defined in `dojo-types` & use it in `torii` later?
```rust
pub const HEXADECIMAL_STR_PREFIX: &str = "0x";
```
Maybe it's just not needed

Comment (tarrencev) [reply crates/torii/graphql/src/query/mod.rs:134]: Nah I think a constant in this case is a bit overkill

Comment (akhercha) [reply crates/torii/graphql/src/query/mod.rs:135]: Updated!

**Pull request merged**